src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java

```diff
@@ -0,0 +1,91 @@
+package org.jenkinsci.plugins.vsphere;
+
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.TaskListener;
+import hudson.model.AsyncPeriodicWork;
+import hudson.slaves.AbstractCloudSlave;
+import hudson.slaves.Cloud;
+import hudson.model.Label;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.vSphereCloud;
+import org.jenkinsci.plugins.vSphereCloudSlaveComputer;
+import org.jenkinsci.plugins.vSphereCloudSlaveTemplate;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * A {@link AsyncPeriodicWork} that reconciles nodes to meet template values.
+ * <p>
+ * The async work will check the number of deployed nodes and provision (or
+ * delete) additional ones to meet template values. The check is happening every
+ * 2 minutes.
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public final class VSphereNodeReconcileWork extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(VSphereNodeReconcileWork.class.getName());
+
+    public VSphereNodeReconcileWork() {
+        super("Vsphere nodes reconciliation");
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return Functions.getIsUnitTest() ? Long.MAX_VALUE : MIN * 2;
+    }
+
+    @Override
+    public void execute(TaskListener listener) {
+        for (Cloud cloud : Jenkins.getActiveInstance().clouds) {
+            if (!(cloud instanceof vSphereCloud)) continue;
+            for (vSphereCloudSlaveTemplate template : ((vSphereCloud) cloud).getTemplates()) {
+                String templateLabel = template.getLabelString();
+                Label label = Label.get(templateLabel);
+
+                int instancesMin = template.getInstancesMin();
+                List<vSphereCloudSlaveComputer> idleNodes = template.getIdleNodes();
+                List<vSphereCloudSlaveComputer> runningNodes = template.getOnlineNodes();
+                // Get max number of nodes that could be provisioned
+                int globalMaxNodes = ((vSphereCloud) cloud).getInstanceCap();
+                int templateMaxNodes = template.getTemplateInstanceCap();
+                int maxNodes = Math.min(globalMaxNodes, templateMaxNodes);
+
+                // if maxNumber is lower than instancesMin, we have to ignore instancesMin
+                int toProvision = Math.min(instancesMin - idleNodes.size(), maxNodes - runningNodes.size());
+                if (toProvision > 0) {
+                    // provision desired number of nodes for this label
+                    LOGGER.log(Level.INFO, "Pre-creating {0} instance(s) for template {1} in cloud {3}",
+                            new Object[] { toProvision, templateLabel, cloud.name });
+                    try {
+                        cloud.provision(label, toProvision);
```

Comment:
The problem here is that `label` isn't neccessarily unique to one template.

What we want here is to provision an instance for a specific template, rather than leaving it to the `provision` method to select a template that matches the specified label.
(FYI the `provision` method will distribute VMs over all templates matching the specified label if those templates have instance caps on them)

e.g. where I work, we have set of GPU-enabled servers on our vSphere cluster, each of which is capable of running N VMs, so we have one template per server with "max instances" = N, but aside from a specification to limit the template to a specific server, the templates are identical.  This means that if we're empty and then get a request for 5 instances that can run on either or two servers, we'll run 3 on one server and 2 on the other.
...although, in this example, all 5 would take time in arriving (as our vSphere cluster takes a while to clone a new VM from a vSphere template) and then take even longer before they're online (as the VMs run Windows and this takes a while to boot up).

If you compare what you're doing here with what was done on the [OpenStack PR](https://github.com/jenkinsci/openstack-cloud-plugin/pull/205/files) you'll see that the provisioning process in the cloud class was refactored to expose a "provision a new agent from this template" method; I'd suggest that we're going to need a similar thing here, otherwise this method will confuse templates and labels.